File: src/main/java/at/tugraz/experiments/ExternalSort.java
package at.tugraz.experiments;

import at.tugraz.runtime.ObjectReader;
import at.tugraz.runtime.ObjectWriter;
import at.tugraz.util.ObjectFileIndex;
import at.tugraz.util.RootData;
import org.apache.log4j.Logger;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;
import java.util.concurrent.ExecutionException;

public class ExternalSort {

	private static Logger logger = Logger.getLogger(ExternalSort.class);

	public static void main(String[] args) throws IOException, InterruptedException, ExecutionException {

		String inDataPath = System.getProperty("inDataPath");
		String outDataPath = System.getProperty("outDataPath");
		String method = System.getProperty("method");
		int chunkSize = Integer.parseInt(System.getProperty("chunkSize"));

		PriorityQueue<ObjectFileIndex> queue = new PriorityQueue<ObjectFileIndex>();
		ObjectReader reader = new ObjectReader(inDataPath, method);
		int fileCount = (int) Math.ceil((double) reader.getRlen() / chunkSize);
		long[] pageObjectCounter = new long[fileCount];
		long[] readFileObject = new long[fileCount];

		int[] pageCounter = new int[fileCount];
		// set all positions to the beginning of files - zero.
		for(int i = 0; i < fileCount; i++) {
			pageCounter[i] = 0;
		}

		for(int i = 0; i < fileCount && i * chunkSize < reader.getRlen(); i++) {
			RootData[] list = reader.readObjects((long) i * chunkSize, chunkSize);
			Arrays.sort(list);

			ObjectWriter writer = new ObjectWriter(outDataPath+"/"+method+"Java-sorted-"+i+".bin", method, list.length);
			writer.writeObjectToFile(list);
			writer.flush();
		}

		// open all of data files and keep them open to read partially.
		List<ObjectReader> readerArray = new ArrayList<>();
		for(int i = 0; i < fileCount; ++i) {
			String sortedFileName = outDataPath+"/"+method+"Java-sorted-"+i+".bin";
			ObjectReader tmpReader = new ObjectReader(sortedFileName, method);
			readerArray.add(tmpReader);
		}

		// reading objects from the first pages and adding them to a priority queue
		for (int i = 0; i < fileCount; i++) {
			int n = readerArray.get(i).getObjectInEachPage().get(0);
			RootData[] rd = readerArray.get(i).readObjects(0, n);
			pageObjectCounter[i] = n;
			readFileObject[i] = n;
			for (RootData rootData : rd) {
				ObjectFileIndex objectFileIndex = new ObjectFileIndex(i, rootData);
				queue.add(objectFileIndex);
			}
		}
		logger.info("Single-Thread External Sort: First page reading is done! ");
		ObjectWriter writer = new ObjectWriter(outDataPath+"/"+method+"Java-sorted.bin", method, reader.getRlen());

		while(!queue.isEmpty()) {
			ObjectFileIndex tmpObjectFileIndex = queue.remove();
			int fileNumber = tmpObjectFileIndex.getFileIndex();

			// reduce the number of objects read from that file.
			pageObjectCounter[fileNumber] = pageObjectCounter[fileNumber] - 1;

			// If needed load more objects from files.
			// if zero load the next page from file and add objects.
			if(pageObjectCounter[fileNumber] == 0 && pageCounter[fileNumber] < readerArray.get(fileNumber).getObjectInEachPage().size() - 1) {

				// add page counter
				pageCounter[fileNumber] = pageCounter[fileNumber] + 1;
				pageObjectCounter[fileNumber] = readerArray.get(fileNumber).getObjectInEachPage().get(pageCounter[fileNumber]);

				RootData[] rd = readerArray.get(fileNumber).readObjects(readFileObject[fileNumber], (int) pageObjectCounter[fileNumber]);
				readFileObject[fileNumber] += rd.length;

				for (RootData rootData : rd) {
					ObjectFileIndex objectFileIndex = new ObjectFileIndex(fileNumber, rootData);
					queue.add(objectFileIndex);
				}
			}
			writer.writeObjectToFile(tmpObjectFileIndex.getValue());
		}
		writer.flush();
	}
}

File: src/main/rust/src/ExternalSort.rs
#![allow(non_snake_case)]
#![allow(dead_code)]
#![allow(unused_variables)]

use std::{io, env};
use crate::runtime::ObjectReader::ObjectReader;
use crate::runtime::ObjectWriter::ObjectWriter;
use crate::tweetStructs::TweetStatus::TweetStatus;
use priority_queue::PriorityQueue;
use std::cmp::Reverse;
use crate::util::ObjectFileIndex::ObjectFileIndex;

mod tweetStructs;
mod runtime;
mod util;

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    let inDataPath: &str = args[1].as_str();
    let outDataPath: &str = args[2].as_str();
    let method: &str = args[3].as_str();
    let chunkSize: u32 = args[4].parse().unwrap();

    let mut queue: PriorityQueue<ObjectFileIndex, Reverse<usize>> = PriorityQueue::new();
    //let mut queue = PriorityQueue::new();
    let mut reader = ObjectReader::new1(inDataPath, method);
    let fileCount = (reader.getRlen() as f32 / chunkSize as f32).ceil() as u32;
    let mut pageObjectCounter: Vec<u32> = vec![0; fileCount as usize];
    let mut readFileObject: Vec<u32> = vec![0; fileCount as usize];
    let mut pageCounter: Vec<u32> = vec![0; fileCount as usize];

    let mut i = 0;
    while i < fileCount && i * chunkSize < reader.getRlen() {
        let mut list: Vec<TweetStatus> = vec![];
        reader.readObjects(i * chunkSize, chunkSize, &mut list);
        list.sort_by(|cu, ot| cu.getOrder().cmp(&ot.getOrder()));

        let mut writer = ObjectWriter::new1(format!("{}/{}Rust-sorted-{}.bin", outDataPath, method, i).as_str(), method, list.len() as u32);
        for tweet in list {
            writer.writeObjectToFile(tweet);
        }
        writer.flush();
        i += 1;
    }

    let mut readerArray: Vec<ObjectReader> = vec![];
    for i in 0..fileCount {
        let sortedFileName = format!("{}/{}Rust-sorted-{}.bin", outDataPath, method, i);
        let reader = ObjectReader::new1(sortedFileName.as_str(), method);
        readerArray.push(reader);
    }

    // reading objects from the first pages and adding them to a priority queue
    for i in 0..fileCount {
        let n = readerArray[i as usize].getObjectInEachPage()[&1];
        let mut rd: Vec<TweetStatus> = vec![];
        readerArray[i as usize].readObjects(0, n, &mut rd);
        pageObjectCounter[i as usize] = n;
        readFileObject[i as usize] = n;
        for tweet in rd {
            let order = tweet.getOrder();
            let objectFileIndex = ObjectFileIndex::new(tweet, i);
            queue.push(objectFileIndex, Reverse(order));
        }
    }
    println!("Single-Thread External Sort: First page reading is done! ");
    let mut writer = ObjectWriter::new1(format!("{}/{}Rust-sorted.bin", outDataPath, method).as_str(), method, reader.getRlen());

    while !queue.is_empty() {
        let tmpObjectFileIndex: ObjectFileIndex = queue.pop().unwrap().0;
        let fileNumber = tmpObjectFileIndex.getFileIndex() as usize;
        // reduce the number of objects read from that file.
        pageObjectCounter[fileNumber] = pageObjectCounter[fileNumber] - 1;

        //If needed load more objects from files.
        //if zero load the next page from file and add objects.
        if pageObjectCounter[fileNumber] == 0 && pageCounter[fileNumber] < readerArray[fileNumber].getObjectInEachPage().len() as u32 - 1 {
            // add page counter
            pageCounter[fileNumber] = pageCounter[fileNumber] + 1;
            pageObjectCounter[fileNumber] = readerArray[fileNumber].getObjectInEachPage()[&pageCounter[fileNumber]];

            let mut rd: Vec<TweetStatus> = vec![];
            let rdLen = readerArray[fileNumber].readObjects(readFileObject[fileNumber], pageObjectCounter[fileNumber], &mut rd);
            readFileObject[fileNumber] += rdLen;

            for tweet in rd {
                let order = tweet.getOrder();
                let objectFileIndex = ObjectFileIndex::new(tweet, fileNumber as u32);
                queue.push(objectFileIndex, Reverse(order));
            }
        }
        writer.writeObjectToFile(tmpObjectFileIndex.getObject());
    }
    writer.flush();
    reader.flush();

    println!("Single-Thread External Sort is Done! ");
    Ok(())
}